### PR TITLE
[1/N] Dynamo skipfiles refactor

### PR DESCRIPTION
**Pull request by yanboliang**
This is 1/N of the dynamo skipfiles/allowed_functions refactor, the major change in this PR includes:
* Refactor & define the [skipfiles rules](https://github.com/pytorch/pytorch/pull/109567/files#diff-5aa3ce9db729bf0901ea97a5d3cc51924cc8575d9c516c1c8f572a35de92544aR56) and interface
* For every ```skipfiles.check```, we return both the check result and the skip/inline reason and log them for debugging.
* We found several latent issues/bugs and incorrect implementations in the codebase, but I'm planning to fix them in follow-up PRs to make the refactor decoupled with bug fixes.
* More details in the inline comments.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/config.py [hunk 1]
```diff
@@ -130,21 +130,8 @@
...
-    torch._refs,
-    torch._prims,
-    torch._decomp,
-    torch.utils._contextlib,
-    torch.utils._pytree,
-    torch.fx._pytree,
-    torch.sparse,
-}
+# legacy config, does nothing now!
+skipfiles_inline_module_allowlist = {}
```
Comment (yanboliang) [inline torch/_dynamo/config.py:134]: I think the skip/inline rules should be defined by developer rather than users, so I moved this to ```skipfiles.py```. If we really need to support users side skip/inline rules, we can extend the current list by adding custom skip/inline lists.

File: torch/_dynamo/skipfiles.py [hunk 2]
```diff
@@ -62,58 +137,39 @@ def _module_dir(m: types.ModuleType):
...
+    _module_dir(torch) + "_functorch/apis.py",
+    _module_dir(torch) + "_functorch/deprecated.py",
+    _module_dir(torch) + "distributed/tensor/parallel/_utils.py",
+    _module_dir(torch) + "distributed/tensor/parallel/style.py",
+    _module_dir(torch) + "distributed/tensor/parallel/_data_parallel_utils.py",
+    _module_dir(torch) + "distributed/_tensor/api.py",
+    _module_dir(torch) + "distributed/_tensor/device_mesh.py",
+    _module_dir(torch) + "distributed/_tensor/placement_types.py",
+    _module_dir(torch) + "distributed/c10d_logger.py",
+    _module_dir(torch) + "distributed/_functional_collectives.py",
```
Comment (yanboliang) [inline torch/_dynamo/skipfiles.py:166]: We should move from hardcode string to ```module.__file__``` way for all files in this list. I'll do it in a followup PR - 2/N, I don't want to mix this refactor with some weird failures together.

File: torch/_dynamo/skipfiles.py [hunk 3]
```diff
@@ -208,51 +255,46 @@ def add(import_name: str):
...
-        return True
-    return bool(SKIP_DIRS_RE.match(filename))
+        return SkipResult(
+            True,
+            "should be skipped according skipfiles.FBCODE_SKIP_DIRS",
+        )
+    if bool(SKIP_DIRS_RE.match(filename)):
+        return SkipResult(True, "should be skipped according skipfiles.SKIP_DIRS")
+    else:
+        return SkipResult(False, "inlining by default")
```
Comment (yanboliang) [inline torch/_dynamo/skipfiles.py:286]: For the ```check``` function, we can eventually remove the ```allow_torch``` argument in the future. There are two places to call this function, 1/ in ```eval_frame.py``` where we start tracing a frame (allow_torch=False); 2/ in ```symbolic_convert.py``` where we plan to inline a function (allow_torch=True). I didn't see too much difference between these two cases. The prototype I did hit several latent bugs in our codebase, so I'll leave it to a separate PR - 3/N.

**Comment by ezyang** [comment]
cc @Kronuz, good opportunity to ask questions on the code

File: torch/_dynamo/allowed_functions.py [hunk 4]
```diff
@@ -119,10 +119,6 @@ def _disallowed_function_ids():
         torch._C._dynamo.eval_frame.unsupported,
         torch.Tensor.__init__,
     ]
-    if torch.distributed.is_available():
-        from torch.distributed import _functional_collectives
-
-        config.skipfiles_inline_module_allowlist.add(_functional_collectives)
```
Comment (ezyang) [inline torch/_dynamo/allowed_functions.py:125]: confirmed MOVED

File: torch/_dynamo/config.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/config.py:134]: confirmed no use sites via fbgs

File: torch/_dynamo/eval_frame.py [hunk 5]
```diff
@@ -186,8 +186,9 @@ def __init__(self, mod: torch.nn.Module, dynamo_ctx):
 
     def _initialize(self):
         # Do this stuff in constructor to lower overhead slightly
-        if isinstance(self._orig_mod.forward, types.MethodType) and skipfiles.check(
-            inspect.getsourcefile(self._orig_mod.forward)
+        if (
+            isinstance(self._orig_mod.forward, types.MethodType)
+            and skipfiles.check(inspect.getsourcefile(self._orig_mod.forward)).skipped
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:191]: This seems like an error prone refactor, since you are now returning a tuple whereas previous it was returning a bool. Are you going to do something like disable `__bool__` on the ... tuple I guess to avoid bugs where you're on the old call site? Or you might consider keeping `skipfiles.check` at the old API and have a `check_verbose` or something.

Comment (ezyang) [reply torch/_dynamo/eval_frame.py:191]: I am not going to audit that you got all the call sites

File: torch/_dynamo/variables/builder.py [hunk 6]
```diff
@@ -522,11 +522,12 @@ def index_source(key):
             )
         elif (
             istype(value, (type, types.FunctionType))
-            and skipfiles.check(getfile(value), allow_torch=True)
+            and skipfiles.check(getfile(value), allow_torch=True).skipped
             and not inspect.getattr_static(value, "_torchdynamo_inline", False)
         ):
             return SkipFilesVariable(
                 value,
+                skipfiles.check(getfile(value)).reason,
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:530]: it's kind of tiresome that you have to call skipfiles.check twice here

File: torch/_dynamo/skipfiles.py [hunk 7]
```diff
@@ -29,31 +29,106 @@
...
 @skip is a helper decorator that can be applied to your function to cause it to be
 included here.
+
+Dynamo skip/inline rules are defined as follows:
+* Inline is the default behavior and will be used unless explicitly skipped.
+* Dynamo has two SKIPLIST: BUILTIN_SKIPLIST and THIRDPARTY_SKIPLIST.
+    * BUILTIN_SKIPLIST contains builtin python modules, such as abc, collections, etc.
+    * THIRDPARTY_SKIPLIST contains common third party libraries, such as numpy, pandas, etc.
+* Functions in these two SKIPLISTs are always skipped, except when they are explicitly
+    put into the two ALLOWLIST: FILENAME_ALLOWLIST and SUBMODULE_ALLOWLIST.
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:62]: It probably would be clearer if you just have an ordered list "to figure out what the behavior is, follow these steps in order". You have things like "except" and "unless" and it makes it hard to figure out what actually has priority.

Comment (ezyang) [reply torch/_dynamo/skipfiles.py:62]: So in particular, ALLOWLIST has precedence over SKIPLIST has precedence over DEFAULT (inline)

File: torch/_dynamo/skipfiles.py [hunk 8]
```diff
@@ -29,31 +29,106 @@
...
+* PyTorch(torch) is in the BUILTIN_SKIPLIST by default, but there are many cases
+    where we want inline the functions under torch namespace. We should add them
+    into FILENAME_ALLOWLIST or SUBMODULE_ALLOWLIST to make dynamo inline those functions.
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:65]: I'm not sure, just reading this, if ALLOWLIST refers to allowed in graph or not, which would imply you don't inline.

File: torch/_dynamo/skipfiles.py [hunk 9]
```diff
@@ -62,58 +137,39 @@ def _module_dir(m: types.ModuleType):
...
     torch.nn.Sequential.__init__.__code__.co_filename,
     torch.set_rng_state.__code__.co_filename,
     torch._inductor.test_operators.__file__,
     torch.utils._content_store.__file__,
-    # These are dynamo files!
     external_utils.__file__,
-    comptime.__file__,  # Want to inline these helpers
+    comptime.__file__,
+    torch.optim._functional.__file__,
+    torch.utils._foreach_utils.__file__,
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:149]: Why not do these consistently with below?

File: torch/_dynamo/eval_frame.py (hunk 5 above)
Comment (yanboliang) [reply torch/_dynamo/eval_frame.py:191]: Yea, this is a bit error prone, I can switch to a new name.

File: torch/_dynamo/skipfiles.py (hunk 7 above)
Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:62]: Right, I'll refactor is as a priority list.

File: torch/_dynamo/skipfiles.py (hunk 8 above)
Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:65]: Haha, actually this is a terminology confusion or misuse. Everything ALLOW in ```skipfiles.py``` means not skip(inline), this is orthogonal with ALLOW in ```allow_functions.py``` which decided if wrapping it as ```TorchVariable```. I was thinking to name the above lists as  ```FORCE_INLINE_FILENAME_LIST``` and ```FORCE_INLINE_SUBMODULE_LIST``` to avoid confusion. Let's rename it.

File: torch/_dynamo/skipfiles.py (hunk 9 above)
Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:149]: Legacy code, we should make them consistent. But I hit some weird error now, so keep as is. Will address later or in a separate PR.

File: torch/_dynamo/config.py (hunk 1 above)
Comment (yanboliang) [reply torch/_dynamo/config.py:134]: Yes

File: torch/_dynamo/allowed_functions.py (hunk 4 above)
Comment (yanboliang) [reply torch/_dynamo/allowed_functions.py:125]: Yes

File: torch/_dynamo/variables/builder.py (hunk 6 above)
Comment (yanboliang) [reply torch/_dynamo/variables/builder.py]: [````Assignment Expressions```](https://peps.python.org/pep-0572/) can help us, but it's available since Python 3.8, we still support 3.7.

File: torch/_dynamo/symbolic_convert.py [hunk 10]
```diff
@@ -2234,9 +2234,8 @@ def check_inlineable(func):
         except NotImplementedError:
             pass  # closures
 
-        if skipfiles.check(
-            func.get_filename()
-        ) and not skipfiles.is_torch_inline_allowed(func.get_filename()):
+        result = skipfiles.check_verbose(func.get_filename(), allow_torch=True)
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:2237]: The before and after here are not obviously equivalent to me

File: torch/_dynamo/eval_frame.py [hunk 11]
```diff
@@ -1221,7 +1224,9 @@ def result_capturing_wrapper(*graph_inputs):
         if (
             (shape_env := getattr(fake_mode, "shape_env", None)) is not None
             and (dim_constraints := shape_env.dim_constraints) is not None
-            and not skipfiles.check(inspect.getsourcefile(call_to_inspect))
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:1224]: you could have just kept the old check() around now LOL

File: torch/_dynamo/skipfiles.py [hunk 12]
```diff
@@ -29,31 +29,114 @@
...
+* If you call functions under skipped modules/files, Dynamo will wrap these functions
+    as SkipFilesVariable. There are a few functions(e.g, collections.OrderedDict) that
+    we have special handling at SkipFilesVariable.call_function.
+
+Overall: *_INLINELIST has precedence over *_SKIPLIST has precedence over DEFAULT (inline)
+
+To figure out what the behavior is, check the following list in order:
+* FILENAME_INLINELIST (Inline if YES)
+* SUBMODULE_INLINELIST (Inline if YES)
+* BUILTIN_SKIPLIST & SUBMODULE_INLINELIST (Skip if YES)
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:75]: you mean SKIPLIST not INLINELIST right

File: torch/_dynamo/skipfiles.py [hunk 13]
```diff
@@ -62,119 +145,90 @@ def _module_dir(m: types.ModuleType):
...
+    torch.distributions,
+    torch.testing,
+    torch.ao.nn,
+    torch._refs,
+    torch._prims,
+    torch._decomp,
+    torch.utils._contextlib,
+    torch.utils._pytree,
+    torch.fx._pytree,
+    torch.sparse,
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:217]: It is nice that you chose not to change observable behavior, that is good.

However, it is still non-obvious to me how to reason out that the new code here does the same thing as the old code (and not something subtly different.) Can you walk us through the changes you made?

File: torch/_dynamo/symbolic_convert.py (hunk 10 above)
Comment (yanboliang) [reply torch/_dynamo/symbolic_convert.py:2237]: You are right, they are not identical! But this is intended, since we want to consolidate these rules. I'm trying to make the skip/inline list and check rules logically more organized in this refactor.

File: torch/_dynamo/skipfiles.py (hunk 12 above)
Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:75]: Oops!

File: torch/_dynamo/skipfiles.py (hunk 13 above)
Comment (yanboliang) [reply torch/_dynamo/skipfiles.py:217]: You are right, they may have subtly different. But I think if we refactor the code, we should correct the mistake we made before, and consolidate these rules into a more organized way, which is the goal of the refactor. I expected there will be slightly difference and trigger some bugs, then we should just fix these bugs using the new structure. I'm happy to walk though these changes and potential changes in my mind.

File: torch/_dynamo/eval_frame.py (hunk 11 above)
Comment (jansel) [reply torch/_dynamo/eval_frame.py:1224]: yeah, let's just define a:
```
def check(f):
   return check_verbose(f).skipped
```
and keep the simpler API

File: benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_inference.csv [hunk 14]
```diff
@@ -31,7 +31,7 @@ maml_omniglot,pass,0
 mnasnet1_0,pass,0
 mobilenet_v2,pass,0
 mobilenet_v3_large,pass,0
-moco,pass,11
+moco,pass,16
```
Comment (jansel) [inline benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_inference.csv:34]: Why these changes?

Comment (yanboliang) [reply benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_inference.csv:34]: I checked these graph breaks logs and found we just repeat the same graph breaks (introduced by distributed all_gather) more times. I'm guessing because of some minor changes on the skip/inline logics, but I didn't get solid evidence yet. I'm thinking ```moco``` is not a very representative model, so just update the expected file. But open to hear your thoughts if this is important.

File: torch/_dynamo/skipfiles.py [hunk 15]
```diff
@@ -29,31 +29,114 @@
...
 Skipfiles (skipped at the file level instead of function level) still apply on a
 frame-by-frame boundary as dynamo traces, but apply to all functions in that file.
 
 @skip is a helper decorator that can be applied to your function to cause it to be
 included here.
+
+Dynamo skip/inline rules & priorities are defined as follows:
+* Inline is the default behavior and will be used unless explicitly skipped.
+* Dynamo has two SKIPLIST: BUILTIN_SKIPLIST and THIRDPARTY_SKIPLIST.
+    * BUILTIN_SKIPLIST contains builtin python modules, such as abc, collections, etc.
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:59]: nit: why don't you mention torch is here lol

File: torch/_dynamo/skipfiles.py [hunk 16]
```diff
@@ -29,31 +29,114 @@
...
+    * THIRDPARTY_SKIPLIST contains common third party libraries, such as numpy, pandas, etc.
+* Functions in these two SKIPLISTs are always skipped, except when they are explicitly
+    put into the two INLINELIST: FILENAME_INLINELIST and SUBMODULE_INLINELIST.
+* PyTorch(torch) is in the BUILTIN_SKIPLIST by default, but there are many cases
+    where we want inline the functions under torch namespace. We should add them
+    into FILENAME_INLINELIST or SUBMODULE_INLINELIST to make dynamo inline those functions.
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:65]: Also, torch will typically not actually induce a graph break when you're dynamo'ing because there's built-in behavior that overrides the skip behavior (TorchVariable)

File: torch/_dynamo/skipfiles.py [hunk 17]
```diff
@@ -29,31 +29,114 @@
...
 """
 A note on skipfiles:
 
-Dynamo consults this file to determine whether code should be compiled or skipped.
+Dynamo consults this file to determine whether function should be inlined or skipped.
 
 A skip applies at the frame boundary, meaning dynamo either triggers a graph break
-at the beginning of the frame or attempts to trace the whole frame.  When skipping
+at the beginning of the frame or attempts to trace/inline the whole frame. When skipping
 a frame, recursively called frames are still traced by dynamo unless also skipped.
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:48]: This might be a good place to mention how skips interact with manually written logic

File: torch/_dynamo/skipfiles.py [hunk 18]
```diff
@@ -62,119 +145,90 @@ def _module_dir(m: types.ModuleType):
...
+    torch.utils._contextlib.__file__,
+    torch.fx._pytree.__file__,
 }
 
 if torch.distributed.is_available():
     # Inline the checkpoint code from distributed
     import torch.distributed.algorithms._checkpoint.checkpoint_wrapper
 
-    FILENAME_ALLOWLIST |= {
+    FILENAME_INLINELIST |= {
```
Comment (ezyang) [inline torch/_dynamo/skipfiles.py:186]: You sure about this? this looks like a semantics change!

Comment (ezyang) [reply torch/_dynamo/skipfiles.py:186]: oh nvm you renamed ALLOWLIST to INLINELIST, carry on

**Review by ezyang — APPROVED** [review]
I did not carefully check that you were semantics preserving, but the new structure looks good. Thanks!

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/skipfiles.py [hunk 19]
```diff
@@ -29,31 +29,114 @@
...
+    we have special handling at SkipFilesVariable.call_function.
+
+Overall: *_INLINELIST has precedence over *_SKIPLIST has precedence over DEFAULT (inline)
+
+To figure out what the behavior is, check the following list in order:
+* FILENAME_INLINELIST (Inline if YES)
+* SUBMODULE_INLINELIST (Inline if YES)
+* BUILTIN_SKIPLIST & THIRDPARTY_SKIPLIST (Skip if YES)
+* Inline by default
+
```
Comment (anijain2305) [inline torch/_dynamo/skipfiles.py:77]: Very nice!

**Review by anijain2305 — APPROVED** [review]

**Comment by yanboliang** [comment]
closing this in favor of #110330
